nextcord/cooldowns/cooldown.py

```diff
@@ -0,0 +1,253 @@
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+
+from __future__ import annotations
+
+import asyncio
+import functools
+import time
+from asyncio import Future
+from asyncio.events import AbstractEventLoop, get_event_loop
+from logging import getLogger
+from typing import TYPE_CHECKING, Callable, Optional, TypeVar
+
+from nextcord.cooldowns import CooldownBucket
+from nextcord.cooldowns.buckets import _HashableArguments
+from nextcord.cooldowns.protocols import Bucket
+from nextcord.exceptions import CallableOnCooldown
+
+logger = getLogger(__name__)
+
+T = TypeVar("T", bound=_HashableArguments)
+
+
+def cooldown(limit: int, per: float, bucket: Optional[Bucket] = None):
+    """
+    A thing
+
+    Parameters
+    ----------
+    limit: int
+        How many call's can be made in the time
+        period specified by ``per``
+    per: float
+        The time period related to ``limit``
+    bucket: Optional[Bucket]
+        The :class:`Bucket` implementation to use
+        as a bucket to separate cooldown buckets.
+
+        Defaults to :class:`CooldownBucket.all`
+
+    Raises
+    ------
+    RuntimeError
+        Expected the decorated function to be a coroutine
+    CallableOnCooldown
+        This call resulted in a cooldown being put into effect
+    """
+    bucket = bucket or CooldownBucket.all
+    _cooldown: Cooldown = Cooldown(limit, per, bucket)
+
+    def decorator(func: Callable) -> Callable:
+        if not asyncio.iscoroutinefunction(func):
+            raise RuntimeError("Expected `func` to be a coroutine")
+
+        attached_cooldowns = getattr(func, "_cooldowns", [])
+        attached_cooldowns.append(_cooldown)
+        setattr(func, "_cooldowns", attached_cooldowns)
+
+        @functools.wraps(func)
+        async def inner(*args, **kwargs):
+            async with _cooldown(*args, **kwargs):
+                result = await func(*args, **kwargs)
+
+            return result
+
+        return inner
+
+    return decorator
+
+
+class _CooldownTimesPer:
+    # Essentially TimesPer but modified
```

Comment:
This should probably be a option in times per / extend it?

Reply:
Also not super fan of the hidden naming here.

Reply:
> This should probably be a option in times per / extend it?

To what end?

Reply:
Not sure what you mean here

Reply:
I don't understand what you wanted from this.

Reply:
I thought it might be easier to modify times_per to have uh a max queue size or similar?

Reply:
There is no queue like the ratelimiter. This will throw a cooldown error if it exceeds the given period

Reply:
N/A